core/system/src/register_gateway.c

```diff
@@ -0,0 +1,127 @@
+ * limitations under the License.
+ */
+#include <uvisor.h>
+#include "context.h"
+#include "halt.h"
+#include "register_gateway.h"
+#include "vmpu.h"
+
+/** Perform the validation of a register gateway.
+ * @internal
+ * @warning This function does not verify that the operation, value and mask
+ * make sense together. The code that performs the operation should check for
+ * this and return an error accordingly.
+ * @param register_gateway[in]  Pointer to the register gateway structure in
+ *                              flash.
+ * @returns the extracted address of the register gateway.
+ */
+static TRegisterGateway * register_gateway_check(TRegisterGateway * register_gateway)
+{
+    /* Verify that the register gateway is in public flash. */
+    /* Note: We only check that the start and end addresses of the gateway are
+     * in public flash. Since the gateway is fairly small, we can safely assume
+     * that all of it is in public flash. */
+    if (!vmpu_public_flash_addr((uint32_t) register_gateway) ||
+        !vmpu_public_flash_addr((uint32_t) ((void *) register_gateway + sizeof(TRegisterGateway) - 1))) {
+            HALT_ERROR(SANITY_CHECK_FAILED, "Register gateway 0x%08X is not in public flash.",
+                       (uint32_t) register_gateway);
+    }
+
+    /* Verify that the register gateway magic is present. */
+    if (register_gateway->magic != UVISOR_REGISTER_GATEWAY_MAGIC) {
+        HALT_ERROR(SANITY_CHECK_FAILED, "Register gateway 0x%08X does not contain a valid magic (0x%08X).",
+                   (uint32_t) register_gateway, register_gateway->magic);
+    }
+
+    /* Verify that the box configuration pointer is in the dedicated flash
+     * region and that it corresponds to the currently active box. */
+    /* Note: We only check that the box pointer is larger than the start of the
+     * pointer region. The subsequent substraction that we do to calculate the
+     * box ID is then guaranteed to be sane. */
+    if (register_gateway->box_ptr < (uint32_t) __uvisor_config.cfgtbl_ptr_start) {
+        HALT_ERROR(SANITY_CHECK_FAILED,
+                   "The pointer (0x%08X) in the register gateway 0x%08X is not a valid box configuration pointer.",
+                   register_gateway->box_ptr, (uint32_t) register_gateway);
+    }
+    uint8_t box_id = (uint8_t) ((uint32_t *) register_gateway->box_ptr - __uvisor_config.cfgtbl_ptr_start);
+    if (box_id != g_active_box) {
+        HALT_ERROR(NOT_ALLOWED, "Register gateway is owned by box %d, while the active box is %d.",
+                   box_id, g_active_box);
+    }
+
+    /* Verify that the address is in one of the allowed memory sections:
+     *   0x40000000 - 0x400FFFFF: Peripheral
+     *   0x42000000 - 0x43FFFFFF: Peripheral alias
+     *   0xE00FF000 - 0xE00FFFFF: Custom ROM Table */
+    uint32_t address = register_gateway->address;
+    if (((address & VMPU_PERIPH_MASK) != VMPU_PERIPH_START) &&
```

Comment:
@Alessandro you can combine both into one rule -and ignore the 1MB hole inbetween.